### DISCUSSION TITLE
Wireguard VPN Client for remote Zigbee device reporting

### DISCUSSION THREAD

**Discussion by Fayrewood** [arendst/Tasmota#20889] [Q&A]
Hi,
I would like to install my Tasmotized Sonoff ZBbridge at our local marina and monitor various Zigbee sensors via MQTT.  

The Sonoff would be simply a guest device on the marina's wifi and there is no chance that they would allow me to set up a site to site VPN on their equipment.  I have Wireguard installed on a server at home and would like the remote devices to report back to my MQTT broker securely using the VPN key exchange.

Is it possible to configure Tasmota to establish a wireguard VPN once it has connected to the public marina wifi?

Many thanks,
Pete

**Comment by sfromis** [comment]
What Tasmota supports instead is [TLS Secured MQTT](https://tasmota.github.io/docs/TLS/).

**Comment by s-hadinger** [comment]
Wireguard is now available: https://github.com/arendst/Tasmota/pull/23347 and https://tasmota.github.io/docs/Wireguard/

**Comment by kalazino** [comment]
Hi everyone,
 I compiled a version of Tasmota for the ESP32C3 with Wireguard enabled, but as soon as I run the WGConnect command, the ESP resets. 
The wireguard.conf configuration is correct, as it works when imported into a Windows client. 
Any suggestions? 
Thanks.

**Reply by s-hadinger** [reply]
Can you share your configuration file (obfuscating the secrets of course)

**Reply by kalazino** [reply]
Hi,
this is the content of wireguard.conf:

[Interface]
PrivateKey = AEKHpFC1R1tsX5HsPj7XXXXXXXXXXXXXXXXXXXXX
Address = 192.168.150.5/32

[Peer]
PublicKey = qCzThtXyOsp8XDXXXXXXXXXXXXXXXXXXXXXXX
AllowedIPs = 192.168.158.254/32
Endpoint = XXXX.sytes.net:13231
PersistentKeepalive = 25

[Tasmota]
Netmask = 255.255.255.0
AutoConnect = 0

In Tasmota console, I get this when I launch the command WGConnect:

13:50:48.666 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":5,"Reason":"Load access fault","EPC":"420db6c4","EXCVADDR":"ec915f6c","CallChain":["420c9e5e","42089b96","420c9aa0","420c9924","420c9e5e","420c9924","420c9f84","42141026","4207e3dc","420c9924","4208f776","420ca0f0","4204ba9a","4207fc52","40388500","42001100","4204bfbe","42001258","4204c03c","42065642","42001238","42006ab0","42002540","42095e01","420ada00","40389c66","42002626","42006cfc","42002626","42012f36","42017c04","4206ecde","4206ec3c","420998ca","4038831e","4206b070","4206b18a","4206ecea","40389c66","4206c900","4207004c"]},"BootCount":13}}

Thanks

**Reply by Noschvie** [reply]
@kalazino to me it looks like the PresharedKey parameter is missing

**Reply by kalazino** [reply]
even setting the presharedkey, the result is the same

**Reply by Noschvie** [reply]
Maybe the same reason as this
https://github.com/arendst/Tasmota/issues/23853

**Reply by kalazino** [reply]
Thanks a lot. In the meantime, I'm testing on an ESP32 S3 and Wireguard is up, but the device often reboots cyclically, no longer reading the Wireguard configuration file... in short, it's not stable. 
I'll try following the instructions in the thread you linked today.
Thanks

**Reply by s-hadinger** [reply]
> Maybe the same reason as this #23853

I doubt, the problem in wireguard was opened before this change

**Reply by s-hadinger** [reply]
I tried to reproduce your problem.

First the configuration you shared should have raised a log error entry at boot:
```
00:00:00.097 WG : WireGuard initialization failed, invalid configuration

# and WgConnect would have raised the following error
09:54:37.492 RSL: RESULT = {"WGConnect":"Not configured"}
```

I added a correct `PresharedKey = <psk>` and it didn't crash.

Can you share you `.map` file produced when compiling your version? With it we can find where the crash happens

**Reply by kalazino** [reply]
Yes, then I added the pre-shared key and it connects. After a few reboots, sometimes it stops connecting and resets. This afternoon I can do some troubleshooting and provide what you requested.

**Reply by s-hadinger** [reply]
Anyway we need the crash report and map file

**Reply by kalazino** [reply]
How do I generate the crash report?

**Reply by s-hadinger** [reply]
It's just this line:
```
13:50:48.666 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":5,"Reason":"Load access fault","EPC":"420db6c4","EXCVADDR":"ec915f6c","CallChain":["420c9e5e","42089b96","420c9aa0","420c9924","420c9e5e","420c9924","420c9f84","42141026","4207e3dc","420c9924","4208f776","420ca0f0","4204ba9a","4207fc52","40388500","42001100","4204bfbe","42001258","4204c03c","42065642","42001238","42006ab0","42002540","42095e01","420ada00","40389c66","42002626","42006cfc","42002626","42012f36","42017c04","4206ecde","4206ec3c","420998ca","4038831e","4206b070","4206b18a","4206ecea","40389c66","4206c900","4207004c"]},"BootCount":13}}
```

**Reply by Noschvie** [reply]
It looks to me like the problem is specific to the C3. The WG connection works with a D1mini, but on the ESP32-C3, it reboots when calling WgConnect in the console.

**Reply by kalazino** [reply]
But even on S3, it's not stable and causes resets. I'll send the requested files this afternoon to try to solve or identify the problem.

**Reply by kalazino** [reply]
@s-hadinger Here is the link to .map:
https://www.transfernow.net/dl/20250830CXlOjvOb

at this time, even though I have done some reboots (unplugging the power) it seems that the problem is not present

**Reply by kalazino** [reply]
Update: Without changing anything, the device has been working normally for 1 full day
